src/random.cr

```diff
@@ -242,21 +242,30 @@ module Random
     {% end %}
   {% end %}
 
-  # Returns a random `Float64` which is greater than or equal to `0`
+  # Returns a random `Float` which is greater than or equal to `0`
   # and less than *max*.
   #
   # ```
   # Random.new.rand(3.5)    # => 2.88938
   # Random.new.rand(10.725) # => 7.70147
   # ```
-  def rand(max : Float) : Float64
+  def rand(max : Float64) : Float64
     unless max > 0
       raise ArgumentError.new "Invalid bound for rand: #{max}"
     end
     max_prec = 1u64 << 53 # Float64, excluding mantissa, has 2^53 values
     rand(max_prec) / max_prec.to_f64 * max
   end
 
+  # :ditto:
+  def rand(max : Float32) : Float32
+    unless max > 0
+      raise ArgumentError.new "Invalid bound for rand: #{max}"
+    end
+    max_prec = 1u32 << 22 # Float32, excluding mantissa, has 2^22 values
```

Comment:
Not very sure about this line. In https://en.wikipedia.org/wiki/Floating-point_arithmetic#Internal_representation, I see that the Significant field of Double is 52 (above for Float64 it was written `1u64 << 53`), and single `23`. So logically, I have written `1u32 << 22`, but I don't have the proof it is correct.

Reply:
A trivial and proofed implementation would be `rand(max.to_f64).to_f32`.

Reply:
That's not idea. We don't have the proof neither for Float64.

Reply:
That's not ideal. We don't have the proof neither for Float64.

Reply:
Regarding the original code: it's certainly not perfect.

The part about "excluding mantissa, has 2^53 values" is relevant only when you're generating a number from 0 to 1. There is very little research done on generating other ranges. Most programming languages go from there and just multiply it by `max` to get other ranges (which is open to all kinds of tiny biases) in just the same way, so we're not worse off.

Why it's good for the range [0, 1]:

https://prng.di.unimi.it/#:~:text=generating%20uniform%20doubles%20in%20the%20unit%20interval

<table><tr><td>

`Float64` can exactly represent all numbers of the form ***k* / 2<sup>53</sup>**, for *k* &leq; 2<sup>53</sup>. (And 53 is the biggest such number where this still holds.)
</td></tr></table>

That's not an exhaustive list of numbers in the range [0, 1] that it can represent (they get denser and denser closer to 0), so this method is not claiming to fill all of them. 

> side fun fact: in the range [1, 2) there *are* actually exactly 2<sup>52</sup> `Float64`s, uniformly distributed, same goes for any range [*a*, 2*a*)
>
> ```console
> $ crystal eval 'a=(1 + 0o31337/(1u64<<52)); p pointerof(a).as(UInt64*).value.to_s(8)'
> "377600000000000031337"
> ```
>
> The number 53 comes from the fact that we can take the range [0.5, 1) which also has 2<sup>52</sup> `Float64`s, then add to it the range [0, 0.5) which has *more* than 2<sup>52</sup> and we don't care about the "more" part.
>
> Background reading: https://fabiensanglard.net/floating_point_visually_explained/

So I was saying that the method doesn't try to fill all possible `float`s. But what it *does* trivially guarantee is that all the numbers that it picks will be equally likely, and that a float can always exactly represent each of those numbers. So there are no biases for the range [0, 1].

I am quite sure that the exact same property would apply to `Float32`:

---

> side fun fact: in the range [1, 2) there *are* actually exactly 2<sup>23</sup> `Float32`s, uniformly distributed, same goes for any range [*a*, 2*a*)
>
> ```console
> $ crystal eval 'a=(1 + 0o31337/(1u64<<23)).to_f32; p pointerof(a).as(UInt32*).value.to_s(8)'
> "7740031337"
> ```

<table><tr><td>

`Float32` can exactly represent all numbers of the form ***k* / 2<sup>24</sup>**, for *k* &leq; 2<sup>24</sup>. (And 24 is the biggest such number where this still holds.)
</td></tr></table>

So indeed I say that writing `24` here is at least as correct as writing `53` in the other method.
(and yes, it's `24`, not `22`)